craft_application/application.py

```diff
@@ -261,6 +273,40 @@ def _resolve_project_path(self, project_dir: pathlib.Path | None) -> pathlib.Pat
 
         return (project_dir / f"{self.app.name}.yaml").resolve(strict=True)
 
+    def get_build_plan(
+        self, yaml_data: dict[str, Any]
+    ) -> Sequence[craft_platforms.BuildInfo] | list[models.BuildInfo]:
+        """Get the build plan for the passed project data.
+
+        :param yaml_data: The raw project data from *craft.yaml, loaded is a dictionary.
+        :returns: a Sequence of BuildInfo objects.
+
+        This returns from the BuildPlannerClass if one is set. Otherwise, the default
+        behaviour is to use the craft-platforms default build planner. Applications
+        that need to override this may implement different build planning functionality
+        here instead.
+        """
+        # Compatibility until the next major version removes the BuildPlanner.
+
+        # We cannot simply check identity because the BuildPlannerClass is actually
+        # a pydantic.Field object if not overridden.
+        if self.app.BuildPlannerClass != None:  # noqa: E711
+            warnings.warn(
+                "The BuildPlanner class is deprecated. Override Application.get_build_plan instead.",
+                category=DeprecationWarning,
+                stacklevel=3,
+            )
+            build_planner = self.app.BuildPlannerClass.from_yaml_data(  # type: ignore[union-attr]
+                yaml_data, self._resolve_project_path(self.project_dir)
+            )
+            return build_planner.get_build_plan()
+
+        return craft_platforms.get_platforms_build_plan(
+            base=cast(str, yaml_data.get("base")),
```

Comment:
What do the validation errors look like now? For example, what gets output if a project file is missing the `base`?

Reply:
That's a good point. I should probably move up the pydantic validation.

Reply:
maybe too late to suggest this but have you considered just having `BuildPlannerClass.get_build_plan()` return a list of craft_platform BuildInfos?